### PR TITLE
parallel nic adding

### PR DESCRIPTION
**Pull request by DaanHoogland**
### Description

This PR checks for parallel calls to addNetToVM and allows for them to be queued per network

Fixes: #5499

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [x] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

manually by adding five nets three times at once I.E
```
for net in \
      9dd1b273-6d72-4602-bf2b-f64fa55c3824 9dd1b273-6d72-4602-bf2b-f64fa55c3824 9dd1b273-6d72-4602-bf2b-f64fa55c3824\
      b26eabb7-5164-4da4-a1fe-7f6100678147 b26eabb7-5164-4da4-a1fe-7f6100678147 b26eabb7-5164-4da4-a1fe-7f6100678147\
      58561085-a7c5-43b7-8f10-8680825f0fc0 58561085-a7c5-43b7-8f10-8680825f0fc0 58561085-a7c5-43b7-8f10-8680825f0fc0\
      01b81bb8-7515-4cac-a51c-9efd0770a9a5 01b81bb8-7515-4cac-a51c-9efd0770a9a5 01b81bb8-7515-4cac-a51c-9efd0770a9a5\
      43f182a3-30a5-43f6-936f-970487804717 43f182a3-30a5-43f6-936f-970487804717 43f182a3-30a5-43f6-936f-970487804717
do
    cmk add nictovirtualmachine networkid=$net virtualmachineid=557e4c83-09da-4cf6-b5e0-c3ffa8122451 &
done
```

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 1]
```diff
@@ -5618,35 +5631,76 @@ private void checkConcurrentJobsPerDatastoreThreshhold(final StoragePool destPoo
...
+            VirtualMachine vm,
+            Network network,
+            NicProfile requested,
+            CallContext context,
+            User user,
+            Account account,
+            List<VmWorkJobVO> pendingWorkJobs) {
+        VmWorkJobVO workJob = null;
+        for (VmWorkJobVO job : pendingWorkJobs) {
+            if (network.getUuid().equals(job.getSecondaryObjectIdentifier())) {
```
Comment (weizhouapache) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:5664]: @DaanHoogland 
do it need to check network uuid here ?
you have already list vm work jobs by vmid and network uuid.
```
        final List<VmWorkJobVO> pendingWorkJobs = _workJobDao.listPendingWorkJobs(
                VirtualMachine.Type.Instance, vm.getId(),
                VmWorkAddVmToNetwork.class.getName(), network.getUuid());
```

Comment (DaanHoogland) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:5664]: yes @Wei. good point, i did this check first and never removed it. We could remove it now but I will wait till the issue of adding the same network multiple time in parallel is out of the way.

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 2]
```diff
@@ -4021,10 +4021,23 @@ public NicProfile addVmToNetwork(final VirtualMachine vm, final Network network,
         }
     }
 
+    /**
+     * duplicated in {@see UserVmManagerImpl} for a {@see UserVmVO}
+     */
+    private void checkIfNetExistsForVM(VirtualMachine virtualMachine, Network network) {
+        List<NicVO> allNics = _nicsDao.listByVmId(virtualMachine.getId());
```
Comment (nvazquez) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:4028]: Do we need an additional check for the nics state?

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 3]
```diff
@@ -4021,10 +4021,23 @@ public NicProfile addVmToNetwork(final VirtualMachine vm, final Network network,
         }
     }
 
+    /**
+     * duplicated in {@see UserVmManagerImpl} for a {@see UserVmVO}
+     */
+    private void checkIfNetExistsForVM(VirtualMachine virtualMachine, Network network) {
```
Comment (nvazquez) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:4027]: Minor: is there a place we can make this code accessible to both classes?

Comment (sureshanaparti) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:4027]:
```suggestion
    private void checkIfNetworkExistsForVM(VirtualMachine virtualMachine, Network network) {
```

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 4]
```diff
@@ -5618,35 +5631,76 @@ private void checkConcurrentJobsPerDatastoreThreshhold(final StoragePool destPoo
...
                 VirtualMachine.Type.Instance, vm.getId(),
-                VmWorkAddVmToNetwork.class.getName());
+                VmWorkAddVmToNetwork.class.getName(), network.getUuid());
 
         VmWorkJobVO workJob = null;
         if (pendingWorkJobs != null && pendingWorkJobs.size() > 0) {
-            assert pendingWorkJobs.size() == 1;
-            workJob = pendingWorkJobs.get(0);
+            if (pendingWorkJobs.size() > 0) {
+                s_logger.warn(String.format("number of jobs to add net %s to vm %s are %d", network.getUuid(), vm.getInstanceName(), pendingWorkJobs.size()));
```
Comment (sureshanaparti) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:5639]:
```suggestion
                s_logger.warn(String.format("Number of pending jobs to add network %s to vm %s are %d", network.getName(), vm.getInstanceName(), pendingWorkJobs.size()));
```

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 2 above)
Comment (DaanHoogland) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:4028]: not sure what you mean, there may be only one nic per network per VM. Does the state matter, @nvazquez ?

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 3 above)
Comment (DaanHoogland) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:4027]: I would like to look for that indeed, but there is still a secondary (corner-case) problem that needs fixing. Hence this intermediate solution. I'm thinking of a service in `QueryManager`

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 4 above)
Comment (DaanHoogland) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:5639]: @suresh, `net.name` may be duplicate and less helpfull, don't you think? (rest of your sugestions I will apply when we agree)

**Comment by DaanHoogland** [comment]
I noticed snags with this code. I think it's not far from being a solution, but still need to add some code :( , @nvazquez

**Comment by DaanHoogland** [comment]
seems to work now
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by sureshanaparti** [comment]
@blueorangutan test matrix

**Review by weizhouapache — APPROVED** [review]
manually tested ok.

tried to add 10 L2 networks to a Running vm, all succeed (tested 5 times).

the issue in my testing is, the deviceid of nics are not as same as the order of `networkid` in the api calls.

@div8cn my suggestion is, wait very few seconds (for example 0.2 or 1 seconds) between two api calls.

**Review by nvazquez — APPROVED** [review]
LGTM

**Pull request merged**